Updated ffish.js to 0.4.2

QueensGambit opened the pull request:
Hi @ianfab  and @gbtami ,

I implemented the remaining feature request from #177.

> another missing feature would be important to add, make moveStack of the Board available.
* [x] board.moveStack() returns the move stack in UCI notation.

+ uddated **ffish.js** and **ffish-es6** to **0.3.0**.

gbtami commented:
@QueensGambit one minor notice. Instead of `/home/queensgambit/Desktop/Programming/forks/Fairy-Stockfish/tests/variants.ini` you can use src/variants.ini to test `ffish.loadVariantConfig()`
Great work, thx a lot!

QueensGambit commented:
* [x] Load variant configurations as string (instead of a file path) as @ianfab suggested in #182 (comment)
I added the functionality for loading a configuration file.
* **ffish.loadVariantConfig(variantsConfig)** it loads a custom variant configuration from a string
* **ffish.variants()** it returns all currently available variants

The parsing seems to work because, I receive the following stdout-message.
`Parsing variant: 3check-crazyhouse
Parsing variant: coffeehouse
Parsing variant: makrukhouse
Parsing variant: xiangqihouse
Parsing variant: janggihouse
Parsing variant: anti-losalamos
Parsing variant: indiangreat
Parsing variant: peasant
Parsing variant: tictactoe
Parsing variant: connect4
Parsing variant: reversi
Parsing variant: othello
Parsing variant: grandhouse
Parsing variant: shogun
Parsing variant: orda`

However, I'm not sure if the variants were activated because *ffish.variants()** didn't update after calling **ffish.loadVariantConfig(variantsConfig)**.

* [x] Add a new function to iterate over SAN move list using pushSan() (Give move list as space delimited string as parameter)
* **board.pushSanMoves(sanMoves, notation)** it pushes multiple san moves on the board, passed as a string with ' ' as delimiter
* **board.pushMoves(uciMoves)**  it pushes multiple uci-moves on the board, passed as a string with ' ' as delimiter

Moreover, I added a basic method for loading pgn files.
* **ffish.readGamePGN(pgn)** it reads a pgn string and returns a game object
I added three pgn-files (**'deep_blue_kasparov_1997.pgn', 'lichess_pgn_2018.12.21_JannLee_vs_CrazyAra.j9eQS4TF.pgn', 'c60_ruy_lopez.pgn'**) to test if the pgn files a being parsed properly.
The code for handling various pgn is getting complex very quickly. Maybe, I'll refactor it and make it more robust later.
For other variants (e.g. Xiangqi, Shogi), additional methods would be needed.

The header can be retrieved and the moves are available via **game.mainlineMoves()**.
I was unable to add **game.board()** because the _Position_ class of _Fairy-Stockfish_ doesn't allow a copy-constructor.

* **game.headerKeys()** it returns all available header keys of the loaded game
* **game.headers(key)** it returns the value for a given header key of a loaded game
* **game.mainlineMoves()** it returns the mainline of the loaded game in UCI notation

gbtami commented:
@QueensGambit ffish.variants() didn't change after calling ffish.loadVariantConfig(variantsConfig) because of
`variants.parse_istream<true>(ss);`
this should be 
`variants.parse_istream<false>(ss);`
because of if `DoCheck` is `true` checked variants will be deleted at the end.

Another issue is I have to create (and delete) a dummy `Board()` now before I can call `loadVariantConfig()` unless sfInitialized will never be true. Shouldn't `initialize_stockfish()` be independent from Board creation?

ianfab commented:
Thanks. As gbtami pointed out, `DoCheck` should be false when loading variants, as this parameter was introduced in order to be able to check the parsing result without any side effect, which is why the variants are removed again in the end. Maybe the name is slightly misleading, since it enables both validation as well as dry run.

QueensGambit commented:
`DoCheck` is now set to `false` and `load_variant_config()` now initializes SF if not done already.

ianfab commented on src/ffishjs.cpp:
```diff
@@ -41,11 +41,12 @@
   UCI::init(Options);
-  PSQT::init(variants.find(uciVariant)->second);
+  for (std::string uciVariant : variants.get_keys())
+    PSQT::init(variants.find(uciVariant)->second);
```
The initialization of PSQT always overwrites the current values, i.e., only one variant can be active at a time. Therefore the for loop does not make much sense here, but since PSQT anyway are not relevant for ffish.js, it should not have any impact either.

QueensGambit replied:
Makes sense. I removed the PSQT initialisation

gbtami commented:
@QueensGambit there is something problematic using `Board::sfInitialized` in `load_variant_config()`
At least if I change the order of test (move the latest loadVariantConfig test to the first place, before Board creation test) tests will fail:
```
tamas@tami:~/Fairy-Stockfish-QueensGambit/tests/js$ npm test

> ffish@0.4.1 test /home/tamas/Fairy-Stockfish-QueensGambit/tests/js
> mocha



  ffish.loadVariantConfig()
    ✓ it loads a custom variant configuration from a string

  Board()
    ✓ it creates a chess board from the default position

  Board(uciVariant) 
    ✓ it creates a board object from a given UCI-variant

  Board(uciVariant) with large board
    ✓ it creates a large-board object from a given UCI-variant

  Board(uciVariant, fen) 
    ✓ it creates a board object for a given UCI-variant with a given FEN

  Board(uciVariant, fen, is960)
    ✓ it creates a board object for a given UCI-variant with a given FEN and is960 identifier

  board.legalMoves()
    ✓ it returns all legal moves in uci notation as a concatenated string

  board.legalMovesSan()
    1) it returns all legal moves in SAN notation as a concatenated string

  board.numberLegalMoves()
    ✓ it returns all legal moves in uci notation as a concatenated string

  board.push(uciMove)
    ✓ it pushes a move in uci notation to the board

  board.pushSan(sanMove)
    ✓ it pushes a move in san notation to the board

  board.pushSan(sanMove, notation)
    ✓ it pushes a move in san notation to the board

  board.pop()
    ✓ it undos the last move

  board.reset()
    ✓ it resets the board to its starting position

  board.is960()
    ✓ it checks if the board originates from a 960 position

  board.fen()
    ✓ it returns the current position in fen format

  board.setFen(fen)
    ✓ it sets a custom position via fen

  board.sanMove()
    ✓ it converts an uci move into san

  board.sanMove(ffish.Notation)
    ✓ it converts an uci move into san using a given notation

  board.variationSan(uciMoves)
    ✓ it converts a list of uci moves into san notation. The board will not changed by this method.

  board.variationSan(uciMoves, notation)
    ✓ it converts a list of uci moves into san notation given a notation format

  board.variationSan(uciMoves, notation, moveNumbers)
    ✓ it converts a list of uci moves into san notation given a notation format and optionally disabling move numbers

  board.turn()
    ✓ it returns the side to move

  board.fullmoveNumber()
    ✓ it returns the move number starting with 1 and is increment after each move of the 2nd player

  board.halfmoveClock()
    ✓ it returns the halfmoveClock / 50-move-rule-counter

  board.gamePly()
    ✓ it returns the current game ply

  board.isGameOver()
    ✓ it checks if the game is over based on the number of legal moves

  board.isCheck()
    ✓ it checks if a player is in check

  board.isBikjang()
    ✓ it checks if a player is in bikjang (only relevant for janggi)

  board.moveStack()
    ✓ it returns the move stack in UCI notation

  board.pushMoves(uciMoves)
    ✓ it pushes multiple uci moves on the board, passed as a string with ' ' as delimiter

  board.pushSanMoves(sanMoves)
    ✓ it pushes multiple san moves on the board, passed as a string with ' ' as delimiter

  board.pushSanMoves(sanMoves, notation)
    ✓ it pushes multiple san moves on the board, passed as a string with ' ' as delimiter

  ffish.info()
    ✓ it returns the version of the Fairy-Stockfish binary

  ffish.setOption(name, value)
Unable to open file variants.ini
    ✓ it sets a string uci option value pair

  ffish.setOptionInt(name, value)
    ✓ it sets a int uci option value pair

  ffish.setOptionBool(name, value)
    ✓ it sets a boolean uci option value pair

  ffish.readGamePGN(pgn)
    ✓ it reads a pgn string and returns a game object

  game.headerKeys()
    ✓ it returns all available header keys of the loaded game

  game.headers(key)
    ✓ it returns the value for a given header key of a loaded game

  game.mainlineMoves()
    ✓ it returns the mainline of the loaded game in UCI notation

  ffish.variants()
    ✓ it returns all currently available variants


  41 passing (1s)
  1 failing

  1) board.legalMovesSan()
       it returns all legal moves in SAN notation as a concatenated string:
     RuntimeError: memory access out of bounds
      at wasm-function[147]:160
      at wasm-function[203]:2453
      at wasm-function[968]:134
      at wasm-function[342]:58
      at wasm-function[1129]:7
      at Module.dynCall_iii (ffish.js:1:122433)
      at dynCall_iii_12 (eval at makeDynCaller (ffish.js:1:87489), <anonymous>:4:12)
      at Board$legalMovesSan [as legalMovesSan] (eval at new_ (ffish.js:1:92824), <anonymous>:8:10)
      at Context.it (test.js:95:23)



npm ERR! Test failed.  See above for more details.
tamas@tami:~/Fairy-Stockfish-QueensGambit/tests/js$
```

QueensGambit commented:
Hi, I think `ffish.loadVariantConfig()` or the ordering of `ffish.loadVariantConfig()` isn't the root cause of this issue.
I compiled **ffish.js** with `ASSERTIONS=1 -s SAFE_HEAP=1` as suggested by @kripken (https://github.com/emscripten-core/emscripten/issues/5161#issuecomment-297196266).

Now a number of segmentation errors are reported for multiple tests.

```python
> ffish-es6@0.4.1 Fairy-Stockfish/tests/js
> mocha



  Board()
    ✓ it creates a chess board from the default position (364ms)

  Board(uciVariant) 
    ✓ it creates a board object from a given UCI-variant

  Board(uciVariant) with large board
    ✓ it creates a large-board object from a given UCI-variant

  Board(uciVariant, fen) 
    ✓ it creates a board object for a given UCI-variant with a given FEN

  Board(uciVariant, fen, is960)
    ✓ it creates a board object for a given UCI-variant with a given FEN and is960 identifier

  board.legalMoves()
    ✓ it returns all legal moves in uci notation as a concatenated string

  board.legalMovesSan()
segmentation fault
segmentation fault
    1) it returns all legal moves in SAN notation as a concatenated string

  board.numberLegalMoves()
    ✓ it returns all legal moves in uci notation as a concatenated string

  board.push(uciMove)
segmentation fault
segmentation fault
    2) it pushes a move in uci notation to the board

  board.pushSan(sanMove)
segmentation fault
segmentation fault
    3) it pushes a move in san notation to the board

  board.pushSan(sanMove, notation)
segmentation fault
segmentation fault
    4) it pushes a move in san notation to the board

  board.pop()
segmentation fault
segmentation fault
    5) it undos the last move

  board.reset()
segmentation fault
segmentation fault
    6) it resets the board to its starting position

  board.is960()
    ✓ it checks if the board originates from a 960 position

  board.fen()
    ✓ it returns the current position in fen format

  board.setFen(fen)
    ✓ it sets a custom position via fen

  board.sanMove()
    ✓ it converts an uci move into san

  board.sanMove(ffish.Notation)
    ✓ it converts an uci move into san using a given notation

  board.variationSan(uciMoves)
segmentation fault
segmentation fault
    7) it converts a list of uci moves into san notation. The board will not changed by this method.

  board.variationSan(uciMoves, notation)
segmentation fault
segmentation fault
    8) it converts a list of uci moves into san notation given a notation format

  board.variationSan(uciMoves, notation, moveNumbers)
segmentation fault
segmentation fault
    9) it converts a list of uci moves into san notation given a notation format and optionally disabling move numbers

  board.turn()
segmentation fault
segmentation fault
    10) it returns the side to move

  board.fullmoveNumber()
segmentation fault
segmentation fault
    11) it returns the move number starting with 1 and is increment after each move of the 2nd player

  board.halfmoveClock()
segmentation fault
segmentation fault
    12) it returns the halfmoveClock / 50-move-rule-counter

  board.gamePly()
segmentation fault
segmentation fault
    13) it returns the current game ply

  board.isGameOver()
segmentation fault
segmentation fault
    14) it checks if the game is over based on the number of legal moves

  board.isCheck()
segmentation fault
segmentation fault
    15) it checks if a player is in check

  board.isBikjang()
    ✓ it checks if a player is in bikjang (only relevant for janggi)

  board.moveStack()
segmentation fault
segmentation fault
    16) it returns the move stack in UCI notation

  board.pushMoves(uciMoves)
segmentation fault
segmentation fault
    17) it pushes multiple uci moves on the board, passed as a string with ' ' as delimiter

  board.pushSanMoves(sanMoves)
segmentation fault
segmentation fault
    18) it pushes multiple san moves on the board, passed as a string with ' ' as delimiter

  board.pushSanMoves(sanMoves, notation)
segmentation fault
segmentation fault
    19) it pushes multiple san moves on the board, passed as a string with ' ' as delimiter

  ffish.info()
    ✓ it returns the version of the Fairy-Stockfish binary

  ffish.setOption(name, value)
Unable to open file variants.ini
    ✓ it sets a string uci option value pair

  ffish.setOptionInt(name, value)
    ✓ it sets a int uci option value pair

  ffish.setOptionBool(name, value)
    ✓ it sets a boolean uci option value pair

  ffish.readGamePGN(pgn)
    ✓ it reads a pgn string and returns a game object

  game.headerKeys()
    ✓ it returns all available header keys of the loaded game

  game.headers(key)
    ✓ it returns the value for a given header key of a loaded game
segmentation fault
segmentation fault
    20) it returns the value for a given header key of a loaded game
```

There also some open issues regarding out of bounds memory access for emscripten (https://github.com/emscripten-core/emscripten/issues/11208) but it's more likely to be due to segmentation faults.

QueensGambit commented:
I was able to track down the first seg. fault in **move_to_san()** where **pos.do_move()** is called:
https://github.com/ianfab/Fairy-Stockfish/blob/master/src/apiutil.h#L269

The seg. fault occurs by calling **thisThread->nodes.fetch_add(1, std::memory_order_relaxed);** for an uninitialized thread object.
https://github.com/ianfab/Fairy-Stockfish/blob/master/src/position.cpp#L1192

gbtami commented:
Huh, seems using more init() from main.cpp in pyffish prevents such issues in the Python binding.
 https://github.com/ianfab/Fairy-Stockfish/blob/master/src/pyffish.cpp#L343

pyffish doesn't utilize Fairy's search. Credit goes to jromang work https://github.com/jromang/Stockfish/blob/pyfish/src/pyfish.cpp

QueensGambit commented:
Adding the threading initialization code requires `-s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=1` as additional compilation commands.

After compilation I receive another memory error.
```python
> mocha

requested a shared WebAssembly.Memory but the returned buffer is not a SharedArrayBuffer, indicating that while the browser has SharedArrayBuffer it does not have WebAssembly threads support - you may need to set a flag
(on node you may need: --experimental-wasm-threads --experimental-wasm-bulk-memory and also use a recent version)
  1) "before all" hook in "{root}"
```

I'd like to avoid using threads for **ffish.js** because they are not required, but adding **#ifdef**s is not nice either.

---

It looks like adding an **#ifdef** for 
*  **thisThread->nodes.fetch_add(1, std::memory_order_relaxed);** (https://github.com/ianfab/Fairy-Stockfish/blob/master/src/position.cpp#L1192)

* **prefetch(thisThread->materialTable[st->materialKey]);** (https://github.com/ianfab/Fairy-Stockfish/blob/master/src/position.cpp#L1287)

fixes all seg. faults and also allows calling `ffish.loadVariantConfig()` first.

If you agree, I will add them.

ianfab commented:
Yes, adding the ifdefs should be fine. It would be good though if they can be written in a way that they do not interfere with the WASM port that uses threading (https://github.com/ianfab/stockfish.wasm) when the code gets merged there.

QueensGambit commented:
I added `NO_THREADS` as a custom **#ifdef** which shouldn't affect any other builds.

ianfab commented:
Yes, squashing would be nice, thanks. Otherwise looks good to me.
